### DISCUSSION TITLE
Using a template in another blueprint template

### DISCUSSION THREAD

**Discussion by jakob30061** [gtk-rs/gtk4-rs#1601] [Q&A] [answered]
Coming from web frameworks, i am used to reuse widgets.
How can i use a template in another blueprint file as if they were nativ gtk widgets?

only resources i could find was https://gtk-rs.org/gtk4-rs/stable/latest/book/composite_templates.html

Thanks for your help in advance.

*Code:*
```blp
using Gtk 4.0;
using Adw 1;

template $TasksView : Gtk.Box {
  content: # my content is here but i think not important for that question
}
```

```blp
using Gtk 4.0;
using Adw 1;

template $AppWindow : Adw.ApplicationWindow {
  title: _("Tasks");
  default-width: 1050;
  default-height: 750;
  width-request: 300;
  height-request: 300;

  content: Stack main_stack {
    transition-type: crossfade;

    StackPage {
      name: "processes";
      child: TasksView {                              <- Error here
        id: "custom_widget_processes";
      };
    }

    StackPage {
      name: "tasks";
      child: TasksView {                            <- Error here
        id: "custom_widget_processes-1";
      };
    }
  };
}
```
and the imp.rs for the parent widget:
`TasksView::ensure_type();` doesn't work for me
```rs
  use crate::views::tasks::TasksView;

  #[derive(CompositeTemplate, Default)]
  #[template(file = "src/window.blp")]
  pub struct AppWindow {
    pub settings: OnceCell<Settings>,
  }

  #[glib::object_subclass]
  impl ObjectSubclass for AppWindow {
    const NAME: &'static str = "AppWindow";
    type Type = super::AppWindow;
    type ParentType = adw::ApplicationWindow;

    fn class_init(klass: &mut Self::Class) {
      TasksView::ensure_type();

      klass.bind_template();
    }

    fn instance_init(obj: &InitializingObject<Self>) {
      obj.init_template();
    }
  }

  impl ObjectImpl for AppWindow {}
  impl WidgetImpl for AppWindow {}
  impl WindowImpl for AppWindow {}
  impl AdwWindowImpl for AppWindow {}
  impl ApplicationWindowImpl for AppWindow {}
  impl AdwApplicationWindowImpl for AppWindow {}
```

**Comment by bilelmoussaoui** [comment]
I can't answer this without seeing the content of your `TasksView`

**Reply by jakob30061** [reply]
Thanks for the quick response.
Here is my full repo with the important files listed below.

If i replace the 
```blp
child: TasksView {                              <- Error here
  id: "custom_widget_processes";
};
```
with a gtk widget like Gtk.Box it is working.

[TaskView Imp](https://github.com/jakob30061/task-manager/blob/7834d5095eed63a0f4350c8e2102ef06b95869e8/src/views/tasks/imp.rs)

Where i want to use it: 
[Window.rs](https://github.com/jakob30061/task-manager/blob/7834d5095eed63a0f4350c8e2102ef06b95869e8/src/window.rs)
[window.blp](https://github.com/jakob30061/task-manager/blob/7834d5095eed63a0f4350c8e2102ef06b95869e8/src/window.blp)

**Reply by bilelmoussaoui** [reply, marked as answer]
Per the docs in https://jwestman.pages.gitlab.gnome.org/blueprint-compiler/reference/templates.html, you should be writing `child: $TasksView` no?

**Reply by jakob30061** [reply]
This worked. Sorry for this silly question. I couldn't find this info.
